kubespawner/spawner.py

```diff
@@ -1241,19 +1241,20 @@ def _set_deprecated(name, new_name, version, self, value):
 
     def _expand_user_properties(self, template):
         # Make sure username and servername match the restrictions for DNS labels
-        # Note: '-' is not in safe_chars, as it is being used as escape character
-        safe_chars = set(string.ascii_lowercase + string.digits)
+        safe_chars = set(string.ascii_letters + string.digits)
 
         # Set servername based on whether named-server initialised
         if self.name:
             servername = '-{}'.format(self.name)
-            safe_servername = '-{}'.format(escapism.escape(self.name, safe=safe_chars, escape_char='-').lower())
+            safe_servername = '-{}'.format(
```

Comment:
I think it would be good to leave a comment here saying something like "if you ever run into weird bugs when unescape()-ing the servername `allow_collisions=True` is why". Basically some kind of hint that we decided on purpose to make unscaping fragile as a trade off for nicer server names.

Reply:
I can totally do that, but I can't find call to `unescape` anywhere in the code. Is `unescape()`-ing something that could really happen?

Reply:
I couldn't find any calls either (also not in the projects that I know that use kubespawner). This is more of a reminder for people from the future who come cursing because they started using `unescape()` and it is failing on them in weird ways. A comment here that says we made this decision on purpose and that broken unescpae is "by design" might help them avoid hours of debugging time.

Reply:
allow_collisions isn't used for unescape roundrip, though that is one effect. It's used for safety. Without it, you can have two valid usernames map onto the same value. Take for example:

```
In [8]: safe_chars = string.ascii_letters + string.digits

In [9]: escape('a.b', safe=safe_chars, escape_char='-', allow_collisions=True).lower()
Out[9]: 'a-2eb'

In [10]: escape('a-2eb', safe=safe_chars, escape_char='-', allow_collisions=True).lower()
Out[10]: 'a-2eb'
```

Both `a.b` and `a-2eb` are valid users, but both are going to have the same pod name `jupyter-a-2eb`!
The escape character cannot be considered a safe character because that allows collisions across usernames. This in general makes all usernames that have a character that needs escaping vulnerable to collisions from other users (or other server names). In the right circumstances, this can enable malicious users access to other users' private user data if users can pick their own username.

A more robust solution is to make a less common but still valid character the escape. I don't have the lists of safe characters for various fields in front of me, but maybe we can use `_` or something.

On the other hand, if *all* names are assumed safe, then we can skip escaping altogether and just let the errors raise through for invalid names. This is something we could do in e.g. BinderHub, and defer to the components picking the names to pick safe ones. I think allowing disabling escaping altogether is probably preferable to allowing collisions.

Reply:
We're not escaping usernames here, only server names. The main goal is to avoid having pod or PVC names like `jupyter-notebook-username-foo-2dbar` when the user created a server named `foo-bar`.